Dependencies/xr/Source/ARKit/XR.mm

```
@@ -0,0 +1,462 @@
+                    const half4 mixed = mix(cameraSample, babylonSample, babylonSample.a);
+                    return vector_float4(mixed);
+                }
+            )";
+            
+            id<MTLLibrary> lib = CompileShader(source);
+            id<MTLFunction> vertexFunction = [lib newFunctionWithName:@"vertexShader"];
+            id<MTLFunction> fragmentFunction = [lib newFunctionWithName:@"fragmentShader"];
+
+            // Configure a pipeline descriptor that is used to create a pipeline state.
+            MTLRenderPipelineDescriptor *pipelineStateDescriptor = [[MTLRenderPipelineDescriptor alloc] init];
+            pipelineStateDescriptor.label = @"XR Pipeline";
+            pipelineStateDescriptor.vertexFunction = vertexFunction;
+            pipelineStateDescriptor.fragmentFunction = fragmentFunction;
+            pipelineStateDescriptor.colorAttachments[0].pixelFormat = MetalLayer.pixelFormat;
+
+            pipelineState = [MetalDevice newRenderPipelineStateWithDescriptor:pipelineStateDescriptor
+                                                                     error:&error];
+            if (!pipelineState)
+            {
+                NSLog(@"Failed to create pipeline state: %@", error);
+            }
+            commandQueue = [MetalDevice newCommandQueue];
+            
+            auto scale = UIScreen.mainScreen.scale;
+            viewportSize.x = MetalLayer.visibleRect.size.width * scale;
+            viewportSize.y = MetalLayer.visibleRect.size.height * scale;
+        }
+
+        ~Impl()
+        {
+            [sessionDelegate release];
+            [configuration release];
+            [session release];
+        }
+
+        id<MTLLibrary> CompileShader(const char* source)
+        {
+            NSError* error;
+            id<MTLLibrary> lib = [MetalDevice newLibraryWithSource:@(source) options:nil error:&error];
+            if(NULL != error)
+            {
+				NSLog(@"Shader compilation failed: %s"
```

Should we throw an exception? This is what we do on the ARCore side, seems like it would be good to have both behave the same.

---

Done.